### PR TITLE
Remove hooks for person/person_distinct_id deletion

### LINKED ISSUE #7489: Remove receive hooks for person and person_distinct_id
## Is your feature request related to a problem?

We currently define post-receive hooks for when person or person_distinct_ids get created or deleted which create/delete the appropriate data in clickhouse.

As far as I'm aware this is widely used in tests for person creation, but some application logic might also rely on it as well.

There are cases where this can cause a lot of trouble:
1. It doesn't handle mass deletions well - if you delete e.g. 100 persons, 100 separate mutations get queued in clickhouse, slowing down queries and potentially blocking migrations or other queries.
2. It's not consistent. Plugin server creates most people and does most of the management around person rows. We shouldn't have a separate out-of-sync mechanism in the app.

## Describe the solution you'd like

Remove the post-receive hooks, update tests and handle clickhouse data deletion asynchronously in a celery task or in a bi-weekly cronjob.

## Describe alternatives you've considered

Never deleting data.

## Additional context

cc @guidoiaquinti @yakkomajuri @fuziontech who were helping to debug an incident today.

### PR DESCRIPTION
**Pull request by macobo**
## Changes

Fixes https://github.com/PostHog/posthog/issues/7489 - hooks will only be there for tests.

Note that deleting data from clickhouse w/o a schedule is not ideal but it's also better than what we have currently. Drawing a line here to avoid next outages.

## How did you test this code?

See existing tests :)

### CODE REVIEW

File: ee/clickhouse/models/person.py [hunk 1]
```diff
@@ -20,9 +20,10 @@
 from ee.kafka_client.topics import KAFKA_PERSON, KAFKA_PERSON_UNIQUE_ID
 from posthog.models.person import Person, PersonDistinctId
 from posthog.models.utils import UUIDT
+from posthog.settings import TEST
 from posthog.utils import is_clickhouse_enabled
 
-if is_clickhouse_enabled():
+if is_clickhouse_enabled() and TEST:
```
Comment (guidoiaquinti) [inline ee/clickhouse/models/person.py:26]: why do we need this?

Comment (macobo) [reply ee/clickhouse/models/person.py:26]: Tests rely on these hooks for creating person and person_distinct_id. Porting 60+ test files is not urgent either.

File: ee/clickhouse/models/person.py [hunk 2]
```diff
@@ -20,9 +20,10 @@
...
 
     @receiver(post_save, sender=Person)
     def person_created(sender, instance: Person, created, **kwargs):
```
Comment (hazzadous) [inline ee/clickhouse/models/person.py:29]: non-blocking just curious

The original ticket is about deletion, what's the reason to special case `person_created` and `person_distinct_id_created` as well?

File: ee/clickhouse/models/team.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    "person",
+    "person_distinct_id",
+    "person_distinct_id2",
+    "groups",
+    "cohortpeople",
+    "person_static_cohort",
+]
+
+
+def delete_teams_data(team_ids: List[int]):
```
Comment (hazzadous) [inline ee/clickhouse/models/team.py:21]: non-blocking

I assume these alter table ops are going to be reasonably quick and we don't need to worry about e.g. celery timeouts?

File: ee/clickhouse/models/team.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    logger.info(
+        f"Deleting teams data from clickhouse using background mutations.",
+        team_ids=team_ids,
+        tables=TABLES_TO_DELETE_FROM,
+    )
+    for table in TABLES_TO_DELETE_FROM:
+        sync_execute(
```
Comment (hazzadous) [inline ee/clickhouse/models/team.py:28]: non-blocking

Is there any chance that there are events etc in kafka that will be added post this run? Do we need to handle this case.

File: ee/clickhouse/models/team.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    "person_distinct_id2",
+    "groups",
+    "cohortpeople",
+    "person_static_cohort",
+]
+
+
+def delete_teams_data(team_ids: List[int]):
+    logger.info(
+        f"Deleting teams data from clickhouse using background mutations.",
```
Comment (hazzadous) [inline ee/clickhouse/models/team.py:23]: Worth adding in a log for each table? We can then also get an understanding of how long each takes

File: posthog/api/organization.py [hunk 6]
```diff
@@ -176,6 +177,8 @@ def get_object(self):
     def perform_destroy(self, organization: Organization):
         user = cast(User, self.request.user)
         report_organization_deleted(user, organization)
+        team_ids = [team.pk for team in organization.teams.all()]
+        delete_clickhouse_data.delay(team_ids=team_ids)
```
Comment (hazzadous) [inline posthog/api/organization.py:181]: non-blocking

I guess there's a failure case where we rollback postgres, but have deleted clickhouse data. Probably not a case that we care about too much but thought I'd point out.

File: ee/clickhouse/models/team.py (hunk 5 above)
Comment (macobo) [reply ee/clickhouse/models/team.py:23]: Deletion is async so this will return immediately.

File: ee/clickhouse/models/team.py (hunk 4 above)
Comment (macobo) [reply ee/clickhouse/models/team.py:28]: There is - we do turn off the api keys, but everything that's not yet ingested would be. 

The answer is likely to delete clickhouse data async in batches on a weekend - something I didn't want to go into this PR.

File: ee/clickhouse/models/team.py (hunk 3 above)
Comment (macobo) [reply ee/clickhouse/models/team.py:21]: They will return immediately as this operation is implemented as an async mutation. https://clickhouse.com/docs/en/sql-reference/statements/alter/delete/

File: ee/clickhouse/models/team.py (hunk 5 above)
Comment (hazzadous) [reply ee/clickhouse/models/team.py:23]: Ah ok, makes sense

File: ee/clickhouse/models/person.py (hunk 2 above)
Comment (macobo) [reply ee/clickhouse/models/person.py:29]: These were used more pre-plugin server in ingestion. Nowadays only relics go through these code paths - split_person was the last one that hadn't been updated.

File: ee/clickhouse/models/team.py (hunk 4 above)
Comment (hazzadous) [reply ee/clickhouse/models/team.py:28]: For sure, keep this pr focused on not crashing rather than new features!

**Review by hazzadous — APPROVED** [review]
I think it makes sense.

Btw, do you know which tests are relying on the hooks?

**Comment by macobo** [comment]
> Btw, do you know which tests are relying on the hooks?

Just about everything in ee/ folder that does `Person.objects.create`. So just about everything :)

**Pull request merged**